sql: panic: cannot overwrite value on flat Bytes

```
CREATE TABLE IF NOT EXISTS tab_orig AS
	SELECT
		'2001-01-01'::DATE + g AS _date,
		g::STRING AS _string
	FROM
		generate_series(0, 5) AS g;

SET vectorize = experimental_always;

SELECT
	tab_4._string
FROM
	tab_orig AS tab_2, tab_orig, tab_orig AS tab_4
ORDER BY
	tab_2._date DESC
LIMIT
	62;
```

```
panic: cannot overwrite value on flat Bytes: maxSetIndex=61, setIndex=0, consider using Reset [recovered]
        panic: cannot overwrite value on flat Bytes: maxSetIndex=61, setIndex=0, consider using Reset [recovered]
        panic: panic while executing 1 statements: SELECT _._ FROM _ AS _, _, _ AS _ ORDER BY _._ DESC LIMIT _; caused by cannot overwrite value on flat Bytes: maxSetIndex=61, setIndex=0, consider using Reset

goroutine 971 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000864000, 0x48569a0, 0xc0011c6240, 0x395bec0, 0xc0015a3390)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000864000, 0x48569a0, 0xc0011c6240)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x395bec0, 0xc0015a3390)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/exec/execerror.CatchVectorizedRuntimeError.func1(0xc00123a1c8)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/execerror/error.go:82 +0x720
panic(0x395bec0, 0xc0015a3390)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/col/coldata.(*Bytes).Set(0xc001478aa0, 0x0, 0xc001b9a03e, 0x1, 0xffc2)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/col/coldata/bytes.go:61 +0x274
github.com/cockroachdb/cockroach/pkg/sql/exec.(*BytesVecComparator).set(0xc0012cd4d0, 0x1, 0x0, 0x3e)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/vec_comparators.eg.go:106 +0x1eb
github.com/cockroachdb/cockroach/pkg/sql/exec.(*topKSorter).spool(0xc0016e23c0, 0x48569a0, 0xc0000ab700)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/sorttopk.go:172 +0x602
github.com/cockroachdb/cockroach/pkg/sql/exec.(*topKSorter).Next(0xc0016e23c0, 0x48569a0, 0xc0000ab700, 0x7f05a587ce00, 0x100)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/sorttopk.go:96 +0x79
github.com/cockroachdb/cockroach/pkg/sql/exec.(*simpleProjectOp).Next(0xc0003774c0, 0x48569a0, 0xc0000ab700, 0x400, 0x20300000000000)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/simple_project.go:93 +0x4b
github.com/cockroachdb/cockroach/pkg/sql/exec.(*limitOp).Next(0xc0000ebfc0, 0x48569a0, 0xc0000ab700, 0x7f05a587ce00, 0x20300000000000)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/limit.go:54 +0x55
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).next(0xc000a3d800, 0x726aef, 0x8, 0xc0016dbdc0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:144 +0x40e
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).nextAdapter(...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:135
github.com/cockroachdb/cockroach/pkg/sql/exec/execerror.CatchVectorizedRuntimeError(0xc00123a1d8, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/execerror/error.go:91 +0x55
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).Next(0xc000a3d800, 0xc0000ab700, 0x3f0aea5, 0xc, 0x48569a0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:179 +0x4d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x48569a0, 0xc0000ab700, 0x486fca0, 0xc000a3d800, 0x48309e0, 0xc001246000)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000a3d800, 0x48569a0, 0xc0000ab700)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc001591440, 0x4856a60, 0xc0012ccd80, 0x407f800, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:584 +0x1ec
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0001088c0, 0xc0013e5da0, 0xc001615dd0, 0xc00123a878, 0xc001246000, 0xc000864420, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:334 +0x381
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0001088c0, 0x4856a60, 0xc0011e5350, 0xc000864420, 0xc0013e5da0, 0xc001615dd0, 0x48587e0, 0xc001591200, 0xc001246000, 0xc0008641d8)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:959 +0x20a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000864000, 0x4856a60, 0xc0011e5350, 0xc000864350, 0x3, 0x7f05a7e46f40, 0xc000042f68, 0xc00123aa01, 0x0, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:808 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000864000, 0x4856a60, 0xc0011e5350, 0xc000864350, 0x7f05a7e46f40, 0xc000042f68, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:700 +0x75b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000864000, 0x4856a60, 0xc0011e5350, 0x4869560, 0xc00097af00, 0xc0006e450a, 0x6f, 0x0, 0x3, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:416 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000864000, 0x4856a60, 0xc0011e5350, 0x4869560, 0xc00097af00, 0xc0006e450a, 0x6f, 0x0, 0x3, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:98 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc000864000, 0x48569a0, 0xc00097af40, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1211 +0x1aa2
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000864000, 0x48569a0, 0xc0011c6240, 0xc00024d780, 0x5400, 0x15000, 0xc00024d818, 0xc0003d00d0, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000c60000, 0x48569a0, 0xc0011c6240, 0xc000864000, 0x5400, 0x15000, 0xc00024d818, 0xc0003d00d0, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc00043c1ce, 0xc001304160, 0x48569a0, 0xc0011c6240, 0xc0003d00d0, 0xc000c60000, 0xc000042900, 0x4867ae0, 0xc001304140, 0xc0012d4a80, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:513 +0x17b
```

--- a/pkg/col/coldata/bytes.go
+++ b/pkg/col/coldata/bytes.go
@@ -120,6 +120,8 @@ func (b *Bytes) Slice(start, end int) *Bytes {
 // If we copy src into the beginning of dest, we will have to move "world" so
 // that the result is:
 // result Bytes: "aworld", offsets: []int32{0, 1}, lengths: []int32{1, 5}
+// Similarly, if "a", is instead "alongerstring", "world" would have to be
+// shifted right.
 func (b *Bytes) CopySlice(src *Bytes, destIdx, srcStartIdx, srcEndIdx int) {
 	if destIdx < 0 || destIdx > b.Len() {
 		panic(fmt.Sprintf("dest index %d out of range (len=%d)", destIdx, b.Len()))

--- a/pkg/sql/exec/vec_comparators_tmpl.go
+++ b/pkg/sql/exec/vec_comparators_tmpl.go
@@ -92,8 +92,18 @@ func (c *_TYPEVecComparator) set(srcVecIdx, dstVecIdx int, srcIdx, dstIdx uint16
 		c.nulls[dstVecIdx].SetNull(dstIdx)
 	} else {
 		c.nulls[dstVecIdx].UnsetNull(dstIdx)
+		// {{ if eq .LTyp.String "Bytes" }}
+		// Since flat Bytes cannot be set at arbitrary indices (data needs to be
+		// moved around), we use CopySlice to accept the performance hit.
+		// Specifically, this is a performance hit because we are overwriting the
+		// variable number of bytes in `dstVecIdx`, so we will have to either shift
+		// the bytes after that element left or right, depending on how long the
+		// source bytes slice is. Refer to the CopySlice comment for an example.
+		execgen.COPYSLICE(c.vecs[dstVecIdx], c.vecs[srcVecIdx], int(dstIdx), int(srcIdx), int(srcIdx+1))
+		// {{ else }}
 		v := execgen.UNSAFEGET(c.vecs[srcVecIdx], int(srcIdx))
 		execgen.SET(c.vecs[dstVecIdx], int(dstIdx), v)
+		// {{ end }}
 	}
 }
 

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -636,3 +636,16 @@ SELECT a, ordinality*2 FROM t_38995 WITH ORDINALITY
 1 2
 2 4
 3 6
+
+# Test for #39827, top k sort with bytes.
+statement ok
+CREATE TABLE t_39827 (a STRING)
+
+statement ok
+INSERT INTO t_39827 VALUES ('hello'), ('world'), ('a'), ('foo')
+
+query T
+SELECT a FROM t_39827 ORDER BY a LIMIT 2
+----
+a
+foo

